Hybrid FSE compatibility

abaicus opened the pull request:
### Summary
Adds compatibility for FSE in a hybrid mode. The PR should work with the child theme here: https://github.com/Codeinwp/neve-fse

After activating the child theme, a section should appear in the customizer, marked as "Beta". 

It adds a new Section in the customizer with toggles for the standard templates:
- Index;
- Front Page;
- Archive;
- 404;
- Search;
- Page;
- Single Post;

We provide some basic templates for all these things mentioned above inside the child theme mentioned above.

<!-- Please describe the changes you made. -->

### Will affect visual aspect of the product
<!-- It includes visual changes? -->
NO

### Screenshots <!-- if applicable -->

<details><summary>Customizer Panel GIF</summary>

_Note: Some of the strings in the GIF below have been removed/changed to keep the translation impact at a minimum._

![fse-customize](https://user-images.githubusercontent.com/15010186/210397618-1be4f29f-be80-414a-8e48-afc17b892a44.gif)

</details>


### Test instructions
<!-- Describe how this pull request can be tested. -->

- Enable the child theme here: https://github.com/Codeinwp/neve-fse;
- Enable the FSE integration inside the customizer panel;
- Enable the different templates and check that they are loaded correctly;
- Check that header/footer toggles work fine;
- Check that the site editor shows the templates correctly and they are editable and render properly on the frontend;

### Known issues:
- Sometimes, depending on what templates are enabled and if the site uses a static front page or not, clicking the Site menu item will break the current screen
![image](https://user-images.githubusercontent.com/15010186/210398277-75f4687d-fc8b-49a5-ae31-6c4b0a708f6b.png)


<!-- Issues that this pull request closes. -->
Closes Codeinwp/neve-pro-addon#1583.
<!-- Should look like this: `Closes #1, #2, #3.` . -->

irinelenache commented:
@abaicus Nice work 🚀 Tested the feature and here is what i found:

- [ ] First thing i see in Customizer is the **Use Site Editor** button, but when i click it, it won't work and i'll go back to dashboard https://vertis.d.pr/v/u4Tmnu After enabling 
- [ ] In frontend i still see the header and the footer as they are set in customizer, even if i activated them in the [full site editing section](https://vertis.d.pr/i/uJ33r8) and i also [changed a bit the templates](https://vertis.d.pr/i/jqL7lW)
- [ ] The blog template si not visible [in the editor](https://vertis.d.pr/i/QYhAxN) even if it's activated in customizer
- [ ] I can't change the text color for some items https://vertis.d.pr/i/l26Qpk They inherit the link color and i can change that only for all links at once

You can check the issues here:
```
https://neve-fse.vertisite.cloud/wp-admin
User: depesepowa4772
Password: h1kHz9K7YIVbvNldGR64
```

Please let me know if i misunderstood anything 🙏 

⚠️ Testing is not complete, i checked the features without using Neve pro for now

abaicus commented:
@irinelenache Thanks for pointing these out! 👍🏻 

>  First thing i see in Customizer is the Use Site Editor button, but when i click it, it won't work and i'll go back to dashboard https://vertis.d.pr/v/u4Tmnu After enabling

I removed the notification altogether as it made things more confusing. It would have been really complex to show it contextually, as we must listen to all the added controls changes.

> In frontend i still see the header and the footer as they are set in customizer, even if i activated them in the [full site editing section](https://vertis.d.pr/i/uJ33r8) and i also [changed a bit the templates](https://vertis.d.pr/i/jqL7lW)

The templates themselves don't use the provided header/footer. To disable the theme header/footer, you must toggle them off in the customizer and manually add the template parts in the editor. I took this decision because I think our header builder system provides more value. The toggles refer to the theme's header/footer, not the one from the site editor. 

>  I can't change the text color for some items https://vertis.d.pr/i/l26Qpk They inherit the link color and i can change that only for all links at once

This issue also happens without using the FSE editor. If you add any post title block and enable the link, the color that applies will still be the link color regardless of what you select. I think this is more of a separate bug we can open an issue for. 

---

I also just noticed that google fonts don't seem to enqueue properly in the site editor. I dug around a bit, but didn't manage to find a way to fix this. @HardeepAsrani, do you have any insights on this? I'm guessing that you might've hit the same issue on Otter.

irinelenache commented:
@abaicus Thanks for clarifying 🙏 

Some other things that i found:

- [ ] The **scroll to top** button is not visible in the pages that use FSE templates https://vertis.d.pr/v/Ylptd3
- [ ] I'm getting [this console error](https://vertis.d.pr/i/4qkRar) in the pages that use FSE templates and our header (i used **Speed** starter site)
- [ ] I'm getting a console error when i'm opening the search canvas https://vertis.d.pr/v/cIzESs
- [ ] With Neve FSE child theme activated, the Shop functions don't work anymore https://vertis.d.pr/v/ZrfTAn I don't exactly know if this is intended or not

HardeepAsrani commented:
@abaicus Yea, in FSE the content is loaded inside an iFrame so anything you enqueue in the editor just doesn't work, so we have made this in Otter: https://github.com/Codeinwp/otter-blocks/blob/master/src/blocks/helpers/google-fonts.js

You can see the `attach` method for the most important part which is to just enqueue whatever you want inside the iFrame of the editor.

abaicus commented:
@irinelenache Thanks for finding the issues. I managed to fix what you mentioned.

> With Neve FSE child theme activated, the Shop functions don't work anymore https://vertis.d.pr/v/ZrfTAn I don't exactly know if this is intended or not

I fixed this one in a way. The issue here is that WooCommerce automatically loads its own templates for those pages and replaces them in frontend once you use an FSE-compatible theme - which the child theme is.

For now, I enabled the default WooCommerce templates in this case. The problem is that we don't have a way to hook in our header/footer in that context. Those templates will be using the FSE header template part the theme provides. I think this is ok for an initial version.

We probably have to take a deeper look at how to disable the automatic loading of templates from WooCommerce.

@selul, what do you think about this?

---

Another problem I noticed is that when using PHP templates for pages, the FSE functionality is dropped, so I removed it in those cases. 

---

Just to let you know, what I mentioned regarding the fonts in the site editor won't be fixed in the initial version:

> I also just noticed that google fonts seem to need to enqueue properly in the site editor.

selul commented:
@abaicus LGTM.

AndreeaCristinaRadacina commented:
Hi there!

When enabling a template in the Customizer, I would expect to see the same name in the editor, too; instead, different names appear for certain templates:
<img width="1295" alt="image" src="https://user-images.githubusercontent.com/86592992/214570217-9ac269a7-7c86-4f40-b397-ff16b15b5904.png">
I think it is a bit confusing:  
Single Post  ( Customizer )-> Single ( Editor )
Single Page ( Customizer ) -> Page ( Editor )
Blog ( Customizer ) -> Index ( Editor )
Would a naming convention help?

Also, is there a specific order in which the enabled template from the Customizer appears in the Editor? I noticed that the order in the Editor wouldn't add the newly enabled templates at the bottom of the existing ones.

Let me know your thoughts!

abaicus commented:
@AndreeaCristinaRadacina The names we used in the customizer were used because we didn't want to add new strings to the theme.

We have no control over the order in which the templates appear in the site editor. We could reorder the controls in the customizer, but I wonder if that makes much sense. I didn't feel an obvious discrepancy in how things are ordered while developing this, but I might have grown accustomed to it, and that's why I wasn't bothered.

@selul, do you think we should make sure the customizer controls use the same order we have in the site editor?

selul commented:
The way it is right now is good for this first Beta version.
